projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,280 @@
+2. Implementing a signature integration with the HSM through the Atomic Swap Proposal module.
+
+## Read Only implementation - Phase 1
+[headless-implementation]: #headless-implementation
+The headless application will be responsible for calling the HSM directly through [the `hsm-dinamo` library](https://www.npmjs.com/package/@dinamonetworks/hsm-dinamo) from Dinamo Networks.
+
+The HSM connection authentication credentials will be stored on the config file along with all other settings for the wallet. This will also allow for configuration through docker parameters. ( See the `config.js.docker` file for more information. )
+```js
+// On the config.js file
+hsmHost: "192.168.0.1"
+hsmUsername: "my_username"
+hsmPassword: "my_password"
+```
+
+One note about this is that a connection may be open at the time the end user decides to reload the configuration settings. For now, this will not be allowed as it would increase the complexity of dealing with closing the current connections. Other connection setting approaches are discussed in the "Future Possibilities" section below.
+
+A dedicated `[POST] /hsm/start` endpoint will be available, receiving the expected `wallet-id` and `hsm-key` body parameters. The `hsm-key` indicates which key on the HSM contains the BIP32 information for this wallet. The headless considers this to be a _hardware wallet_ and realizes the first request to the HSM to validate the existence of the key name.
+
+> **Possibility 1: The key is available**<br>
+> The case flow continues
+>
+> **Possibility 2: The key is not available**<br>
+> An error is thrown indicating the HSM has not been properly configured to have a BIP32 `xPriv` on this specific key.
+
+The key being available, the next call requests the wallet's `xPub`. The parsing of `xPub` information received from the HSM can be based on the desktop hardware wallet [`handlePublicKeyData` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/screens/StartHardwareWallet.js#L106-L138).
+
+If the generated `xPub` is not valid, an error is thrown to the user indicating that the designated HSM key does not contain a valid `BIP32 xPriv` and is unsuitable for using as a wallet.
+
+Once the wallet is started successfully, all other operations are executed in a similar fashion to the current headless implementation of a read-only wallet.
+
+## Signature integration - Phase 2
+The signature of HSM inputs will be made through a `Tx Proposal`, using the Atomic Swap API. The `wallet/atomic-swap/get-my-signatures` route will be modified to identify that the wallet is related to an HSM and will send the data to be signed to the *HSM Client*. This can be based on the desktop hardware wallet [`ledger/sendTx()` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/utils/ledger.js#L126-L166).
```

Comment:
Why do you need to use the `atomic-swap` APIs? We have nowadays the API in the `tx-proposal` routes to build a tx, add signatures to a tx. Do you need anything else? I think you will create a new API in the `hsm` router group to get this signatures from the hsm device, but after you can use the ones from the `tx-proposal` routes.

Reply:
I tried to find an endpoint on `tx-proposal` routes that allowed the signature of transactions, but I couldn't find any.

The `p2sh` routes have a way to generate the signatures, but regular `p2pkh` transactions only have routes to add existing signatures that were already calculated to the transaction ( see [headless docs site](https://wallet-headless-api.docs.hathor.network/#operation/createTxProposal) ).

Maybe there is something I'm missing here?

Reply:
We really don't have an endpoint to generated the signatures in the `tx-proposal` routes. I think we should create one instead of using the `atomic-swap` routes for something that isn't an atomic swap.

Reply:
Makes sense. Added the reference and explanation for the new `hsm/get-my-signatures` route on [ac26cfa](https://github.com/HathorNetwork/rfcs/pull/67/commits/ac26cfa5c7a8db84e58eafe98bff808b502c0299)